File: golang/construct-the-minimum-bitwise-array-i/construct-the-minimum-bitwise-array-i.go
package main

import (
	"fmt"
	"reflect"
)

func minBitwiseArray(nums []int) []int {
	ans := make([]int, 0, len(nums))

	for _, n := range nums {
		found := false

		for x := 0; x < n; x++ {
			if (x | (x + 1)) == n {
				ans = append(ans, x)

				found = true
				break
			}
		}
		if !found {
			ans = append(ans, -1)
		}
	}

	return ans
}

func runTest(testNum int, nums []int, expected []int) {
	result := minBitwiseArray(nums)

	status := "FAIL"
	if reflect.DeepEqual(result, expected) {
		status = "PASS"
	}

	fmt.Printf(
		"Test Case %d: Input %v -> Output %v (Expected: %v) -> %s\n",
		testNum, nums, result, expected, status,
	)
}

func main() {
	// Example 1
	runTest(1,
		[]int{2, 3, 5, 7},
		[]int{-1, 1, 4, 3},
	)

	// Example 2
	runTest(2,
		[]int{11, 13, 31},
		[]int{9, 12, 15},
	)

	// Edge Case
	runTest(3,
		[]int{2},
		[]int{-1},
	)
}

File: leetcode/construct-the-minimum-bitwise-array-i.py
class Solution(object):
    def minBitwiseArray(self, nums):

        ans = []

        for n in nums:
            found = False

            for x in range(n):
                if (x | (x + 1)) == n:
                    ans.append(x)
                    found = True
                    break

            if not found:
                ans.append(-1)

        return ans


if __name__ == "__main__":
    solver = Solution()

    def run_test(case_num, nums, expected):
        result = solver.minBitwiseArray(nums)
        status = "PASS" if result == expected else "FAIL"
        print(
            f"Test Case {case_num}: Input {nums} -> Output {result} (Expected: {expected}) -> {status}"
        )

    # Example 1
    # 2: Even prime, result -1
    # 3: 1 | 2 = 3
    # 5: 4 | 5 = 5
    # 7: 3 | 4 = 7
    run_test(1, [2, 3, 5, 7], [-1, 1, 4, 3])

    # Example 2
    # 11 (1011): x=9 (1001) -> 9|10 = 1011 (11)
    # 13 (1101): x=12 (1100) -> 12|13 = 1101 (13)
    # 31 (11111): x=15 (01111) -> 15|16 = 11111 (31)
    run_test(2, [11, 13, 31], [9, 12, 15])

    # Custom Edge Case
    # Only 2
    run_test(3, [2], [-1])
